BUG: raise for invalid dtypes per issue #15520

analyticalmonk opened the pull request:
closes #15520 

 - [x] closes #xxxx
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

analyticalmonk committed:
TST: wrote test representing bug fix result for #15520

analyticalmonk committed:
BUG: Raise when invalid dtype passed to pandas_dtype

Changes made to pandas_dtype() in pandas/core/dtypes/common.py

jreback commented on pandas/core/generic.py:
```diff
@@ -165,12 +166,15 @@ def _validate_dtype(self, dtype):
 
         if dtype is not None:
             dtype = _coerce_to_dtype(dtype)
```
I think can just directly use ``pandas._dtype`` here. In fact I *think* you can replace *all* usages of ``_coerce_to_dtype`` with ``pandas_dtype``. (you might have to back off some if the tests don't pass), but can investigate those.

analyticalmonk replied:
Using `pandas_dtype()' here will mean that [1] will throw an error. This has been used across various tests, and all of them will break. For example, a [number](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/dtypes/test_missing.py#L185) [of](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/dtypes/test_missing.py#L210) [tests](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/dtypes/test_missing.py#L224) [here](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/dtypes/test_missing.py#L48).

[1] `pd.Series(values, dtype=object)`

jreback replied:
not sure what you mean. This is a fairly generic routine. 

```
In [1]: pandas.core.dtypes.common.pandas_dtype(object)
Out[1]: dtype('O')

In [2]: pandas.core.dtypes.common.pandas_dtype('object')
Out[2]: dtype('O')
```

jreback commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -13,6 +14,15 @@
 
 class TestPandasDtype(tm.TestCase):
 
+    # Passing invalid dtype, both as a string or object, must raise TypeError
```
use a list for checking these cases

analyticalmonk replied:
Done.

jreback replied:
if you want to convert to using more pytest style would be graet as well (could be a followup as well), e.g.: http://pandas-docs.github.io/pandas-docs-travis/contributing.html#transitioning-to-pytest

analyticalmonk committed:
TST: maintain list containing dtypes in TestPandasDtype

analyticalmonk committed:
CLN: Replace _coerce_to_dtype() with pandas_dtype()

Changes made to series.py, generic.py in pandas/core and cast.py in pandas/core/dtypes

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -737,5 +737,7 @@ def pandas_dtype(dtype):
             pass
     elif isinstance(dtype, ExtensionDtype):
         return dtype
+    elif np.dtype(dtype).kind == 'O':
```
this needs to be more defensive

```
try:
   dtype = np.dtype(dtype)
except (TypeError, ValueError):
   raise informative msg here
if dtype.kind == 'O':
    ....
return dtype
```

analyticalmonk replied:
`np.dtype` doesn't differentiate between [1] and [2], any thoughts on how to raise errors for [1] without breaking [2]? Since the change I made would raise an error in both cases.

[1]
```
>>> np.dtype(pd.Timestamp)
dtype('O')
```

[2]
```
>>> np.dtype('object')
dtype('O')
```

jreback replied:
let's add a few more tests cases, FYI (e.g. ``np.float64``, 'float64', np.dtype('float64')``, IOW some objects and strings

maybe something like

``
if isinstance(dtype, (np.dtype, compat.string_types)):
    return np.dtype(dtype)
raise TypeError
```

analyticalmonk committed:
TST: added more test-cases for pandas_dtype() test

analyticalmonk committed:
BUG: pandas_dtype() to raise error for invalid dtype per GH15520

This change ensures that no error is raised when type 'object' is passed.

analyticalmonk commented:
@jreback Updated tests as well as pandas_dtype() function as per your suggestions.

analyticalmonk committed:
style fix

analyticalmonk committed:
TST: Added numpy.object_ dtype to valid pandas_dtype list

analyticalmonk committed:
BUG: Added numpy.dtype_ to valid pandas_dtype() type list

analyticalmonk commented:
@jreback Can this PR be reviewed please?

jreback commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -13,6 +14,19 @@
+            with tm.assertRaisesRegexp(TypeError, msg):
+                pandas_dtype(dtype)
+
+        valid_list = [object, 'float64', np.object_, list,
```
add ``np.dtype('object')``; ``list`` is not a valid dtype.

analyticalmonk replied:
> 'list is not a valid dtype'
@jreback That would result in code such as `Series(dtype=list)` raising an error.
Among other things, this [test](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/test_strings.py#L1214) would break.

jreback replied:
that's wrong it should raise
list is not a valid dtype

analyticalmonk replied:
Should I modify this particular test and remove the checks involving `list' dtype?

jreback replied:
yep

add list to the invalid types checking

jreback commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -13,6 +14,19 @@
 
+    # Passing invalid dtype, both as a string or object, must raise TypeError
+    def test_invalid_dtype_error(self):
+        msg = 'not understood'
```
add the issue number as a comment here

jreback commented on pandas/core/generic.py:
```diff
@@ -164,13 +164,15 @@ def _validate_dtype(self, dtype):
 
         if dtype is not None:
-            dtype = _coerce_to_dtype(dtype)
+            # This would raise an error if an invalid dtype was passed
```
comment not necessary

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -731,11 +731,23 @@ def pandas_dtype(dtype):
+            np.dtype(dtype)
+        except (TypeError, ValueError):
+            raise
+        if dtype in [object, np.object_, list]:
```
list is not valid

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -731,11 +731,23 @@ def pandas_dtype(dtype):
             except TypeError:
                 pass
 
+        elif dtype == 'object':
```
catch this lower (where you catch ``object, np.dtype('object')``

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -731,11 +731,23 @@ def pandas_dtype(dtype):
+            raise
+        if dtype in [object, np.object_, list]:
+            return np.dtype(dtype)
+        elif np.dtype(dtype).kind == 'O':
```
doesn't ``np.dtype(dtype).kind == 'O'`` catch ``object, np.object_, 'object'``?

analyticalmonk replied:
@jreback From what I have understood, any invalid dtype (such as `pd.Timestamp`) should raise an error. `np.dtype(invalid_type).kind = 0` for such objects. However, this will also catch some valid `dtypes` such as `object, np.object_ and 'object'` which we safeguard against by catching them earlier and returning `np.dtype(valid_dtype)` in before this condition is evaluated.

jreback replied:
ok add a comment then as this is a tricky case

analyticalmonk committed:
CLN: refactored code related to issue GH15520

analyticalmonk committed:
TST: Added list to invalid dtype

Changes made to:
 - test_empty_str_methods in test_strings.py
 - test_invalid_dtype_error in dtypes/test_common.py

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -737,5 +737,20 @@ def pandas_dtype(dtype):
             pass
     elif isinstance(dtype, ExtensionDtype):
         return dtype
+    else:
```
we don't need this in the else, it IS the else.
then do

```
try:
   npdtype = np.dtype(dtype)
except ...

....

return npdtype
```
```

jreback replied:
so you can replace usages of ``np.dtype(dtype)`` with npdtype

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -1210,7 +1210,6 @@ def test_empty_str_methods(self):
         empty_str = empty = Series(dtype=str)
         empty_int = Series(dtype=int)
         empty_bool = Series(dtype=bool)
-        empty_list = Series(dtype=list)
```
leave all of these tests, replace 

```
empty_list = Series([], dtype=object)
```

jreback replied:
the tests will still work

jreback replied:
actually, this is the *same* as ``Series(dtype=str)``, so just change that one to ``Series(dtype=object)`` and done

jreback reviewed (commented):
small comments, ping when pushed and green.

jreback commented:
this is ok for 0.20.0, pls add a whatsnew note (bug fix section is fine).

analyticalmonk commented:
@jreback Added the whatsnew note.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1510,6 +1510,12 @@ Conversion
 - Bug in the return type of ``pd.unique`` on a ``Categorical``, which was returning an ndarray and not a ``Categorical`` (:issue:`15903`)
 - Bug in ``Index.to_series()`` where the index was not copied (and so mutating later would change the original), (:issue:`15949`)
 
+
```
just put it in conversion, no need to make a different section.

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -1241,19 +1240,11 @@ def test_empty_str_methods(self):
         tm.assert_series_equal(empty_int, empty.str.rfind('a'))
         tm.assert_series_equal(empty_str, empty.str.pad(42))
         tm.assert_series_equal(empty_str, empty.str.center(42))
-        tm.assert_series_equal(empty_list, empty.str.split('a'))
```
can you add back tests with ``empty_str`` so that all of the original functions are convered

analyticalmonk replied:
Added back the tests after looking out for duplicates; all green now.

analyticalmonk committed:
DOC: added whatsnew entry for PR#16047 addressing GH15520

analyticalmonk committed:
BUG: Added 'O' to pandas_dtype's valid list

analyticalmonk committed:
Merge branch 'master' into patch_for_15520

jreback commented:
add a test for the original issue as well

``pd.Series([], name='time',dtype=pd.Timestamp)``
which works (should raise)

analyticalmonk commented:
@jreback Should I add the test in `core/tests/series/test_dtypes.py`?

jreback commented:
put in ``tests/series/test_constructors`` I think (this is an additional test), leave what you have

Commit:
TST: check for invalid dtype for Series constructor per GH15520

analyticalmonk commented:
@jreback Anything else which is missing?

jreback commented:
thanks @analyticalmonk 

nice fixes!

keem em coming!
